[workflows] Build a container for running CI on github actions

tstellar opened the pull request:
Using a container will allow us to have similar testing environments on both the GitHub hosted runners and the self-hosted runners.

tstellar committed:
[workflows] Build a container for running CI on github actions

Using a container will allow us to have similar testing environments
on both the GitHub hosted runners and the self-hosted runners.

boomanaiden154 commented on .github/workflows/build-ci-container.yml:
```diff
@@ -0,0 +1,51 @@
+        working-directory: ./.github/workflows/containers/github-action-ci/
+        run: |
+          # TODO: Ideally, we would give each container build a unique name,
+          # but we have limited space in the container registry, so we can't
```
Do you have a source for this? I believe this falls under Github packages, which should be free for public packages.

It also seems like they don't have any billing setup for the container registry for even enterprise accounts.

Tagging with the short SHA and then also tagging the image as `latest` would probably be good. There should be a way to just push the tag as well without having to push the image twice. This might be something for a follow up patch though.

tstellar replied:
> Do you have a source for this? I believe this falls under Github packages, which should be free for public packages.
> 
> It also seems like they don't have any billing setup for the container registry for even enterprise accounts.
> 

OK, I missed that the limits only applied to private packages.

> Tagging with the short SHA and then also tagging the image as `latest` would probably be good. There should be a way to just push the tag as well without having to push the image twice. This might be something for a follow up patch though.

I will look into this.

boomanaiden154 commented on .github/workflows/build-ci-container.yml:
```diff
@@ -0,0 +1,51 @@
+permissions:
+  contents: read
+
+on:
```
We probably want a `workflow_dispatch` here too for cases like rebuilding the container with the latest packages in the Ubuntu repos if there's something like a security update.

tstellar replied:
I considered doing this, but decided not to, because I didn't want just anyone to be able to rebuild the container.  One option would be to add a workflow_dispatch, but with a user/team check.

boomanaiden154 commented on .github/workflows/containers/github-action-ci/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+# Need nodejs for some of the GitHub actions.
+RUN apt-get update && \
+    apt-get install -y \
+    abi-compliance-checker \
```
I'm assuming one of these packages pulls in a `gcc` based toolchain? It'll be needed for the documentation tests.

tstellar replied:
It must, but it would be good to explicitly add it and probably clang too.

boomanaiden154 reviewed (commented):
Minor comments.

I'm assuming the plan is to migrate the key existing workflows to use the container image after this lands? We might also want to add a couple more dependencies or sub-containers (that use this one as their base image) for specific workflows as some of them (like the code formatting action) install packages through `apt-get` (or could).

Installing python packages into a container would be nice as well for some of the jobs and would shave off a bit of setup time, but that's already mostly negated by the caching.

For building LLVM/running tests, we'll probably need to make some changes to take more explicit control of the toolchain that we're using. For a build-once run many times container, I think a ThinLTO+PGO+BOLTed clang based toolchain would make a lot of sense, but that is irrelevant for this patch.

tstellar commented:
> Minor comments.
> 
> I'm assuming the plan is to migrate the key existing workflows to use the container image after this lands? We might also want to add a couple more dependencies or sub-containers (that use this one as their base image) for specific workflows as some of them (like the code formatting action) install packages through `apt-get` (or could).
> 

I thought about creating sub-containers, but I wasn't sure if all the extra complexity was actually worth it.  What's the advantage of having separate containers for different types of jobs?

> Installing python packages into a container would be nice as well for some of the jobs and would shave off a bit of setup time, but that's already mostly negated by the caching.
> 

I would like to install the python packages too, I wasn't sure the best way to implement this.  Should I copy all the requirements.txt files into the container directory and then add them to the container?

> For building LLVM/running tests, we'll probably need to make some changes to take more explicit control of the toolchain that we're using. For a build-once run many times container, I think a ThinLTO+PGO+BOLTed clang based toolchain would make a lot of sense, but that is irrelevant for this patch.

This is a good idea.  I've also thought about creating a container that has a pre-filled ccache cache to help speed up builds.

boomanaiden154 commented:
> I thought about creating sub-containers, but I wasn't sure if all the extra complexity was actually worth it. What's the advantage of having separate containers for different types of jobs?

The main advantage I can think of is that it keeps the image size smaller as only the dependencies needed for the job are actually pulled in. For python packages, it also allows us to install everything at the system level in the container rather than having to worry about maintaining multiple virtual environments/trying to consolidate python dependencies.

> I would like to install the python packages too, I wasn't sure the best way to implement this. Should I copy all the requirements.txt files into the container directory and then add them to the container?

I would try and keep each set of dependencies in a separate container. Otherwise the best way to do it (that I can think of) is to consolidate into a single `requirements.txt`/pipenv/something for all the CI jobs. Ideally we want to install the exact transitive closure of the dependencies used for development. Installing all the `requirements.txt` wouldn't guarantee this currently.

> This is a good idea. I've also thought about creating a container that has a pre-filled ccache cache to help speed up builds.

Shouldn't Github's `actions/cache` (or some other caching mechanism) automatically download the most recent cache even if the image changes? We don't want to be rebuilding container images that frequently, and I think the pre-filled ccache would quickly become out of date (pretty much as soon as someone touches a major header).

tstellar commented:
@boomanaiden154 A few updates based on your suggestions:

- Rather than having one container that can be used for every job, I created a container that only works with the llvm-project-tests workflows (and probably the release-binaries workflow too).  I'm still not convinced that having different containers for each job is the right approach, but these workflow are the only jobs that actually need a container right now, because I want to move them to the self-hosted runner.  The rest of the jobs are fine on the GitHub runners, so they don't really need to use containers.
- I installed the most recent clang release (with binaries from the release page) into the container and added its install directory to PATH.  I'm planning on enabling PGO/llvm bolt optimizations for the release builds in the next cycle, so using the release binaries will eventually give us a highly optimized clang for the container as you suggested.
- I added unique tags for each new container build.

TODOs:
- Choose a better container name?
- Add workflow_dispatch.

boomanaiden154 commented on .github/workflows/containers/github-action-ci/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+RUN mkdir -p $LLVM_SYSROOT/bin/ $LLVM_SYSROOT/lib/
+
+RUN curl -O -L https://github.com/llvm/llvm-project/releases/download/llvmorg-$LLVM_VERSION/$LLVM_FILENAME
+#COPY $LLVM_FILENAME .
```
This looks leftover?

boomanaiden154 commented on .github/workflows/containers/github-action-ci/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+ENV PATH=${LLVM_SYSROOT}/bin:${PATH}
+
+# Test clang
+RUN printf '#include <iostream>\nint main(int argc, char **argv) { std::cout << "Hello\\n"; }' | clang++ -x c++ - && ./a.out | grep Hello
```
If we're doing smoke tests on the container, we should do them post-installation.

tstellar replied:
I moved the tests into the workflow file, let me know what you think.

boomanaiden154 replied:
LGTM. Thanks!

boomanaiden154 commented on .github/workflows/build-ci-container.yml:
```diff
@@ -0,0 +1,52 @@
+          echo "container-name-tag=$container_name:$tag" >> $GITHUB_OUTPUT
+
+      - name: Checkout LLVM
+        uses: actions/checkout@v4
```
If we only need the one directory for the step below, maybe we could do a sparse checkout? That would speed things up significantly from my testing.

boomanaiden154 reviewed (approved):
> Rather than having one container that can be used for every job, I created a container that only works with the llvm-project-tests workflows (and probably the release-binaries workflow too). I'm still not convinced that having different containers for each job is the right approach, but these workflow are the only jobs that actually need a container right now, because I want to move them to the self-hosted runner. The rest of the jobs are fine on the GitHub runners, so they don't really need to use containers.

That works. My point was mainly for jobs with extensive python dependencies, but we can look at that later, and that's also only if we want to move those jobs to containers given that they are cheap enough to run on the github hosted runners.

> I installed the most recent clang release (with binaries from the release page) into the container and added its install directory to PATH. I'm planning on enabling PGO/llvm bolt optimizations for the release builds in the next cycle, so using the release binaries will eventually give us a highly optimized clang for the container as you suggested.

Sounds good.

> I added unique tags for each new container build.

Awesome!

Everything in this patch pretty much LGTM. Just minor nits and I don't want to keep anything blocked for future work.

tstellar committed:
Review fixes
